File: A5/bTree.c
/*
 ============================================================================
 Name        : bTree.c
 Author      : Woojin Jang (Student#: 260762536)
 Version     :
 Copyright   : Your copyright notice
 	 	 	 : This code contain codes from Prof. Frank ferrie and TA Katrina Poulin
 Description : Hello World in C, Ansi-style
 ============================================================================
 */

#include <stdio.h>
#include <stdlib.h>
#include "bTree.h"
#include <string.h>


// instance variables
bNode *root = NULL;  // bTree root node
int numNodes = 0;	 // Number of nodes in tree

// methods:
void initTree(){ // this method sets root to null and numNodes to 0
	root = NULL;
	numNodes= 0;
}


void addNode (char * data){
	// Adds a new node to the bTree.
	bNode *current = NULL;
			// empty tree
			if (root == NULL) {
				root = makeNode(data);
			}
			// not empty
			else {
				current = root;
				while (1) {
					// the return value is negative
					// which means the name (data) comes first compare to the root
					if (keyCompareToIgnoreCase(data, current->data) < 0) {
						// left pointer is empty
						if (current->left == NULL) {
							current->left = makeNode(data);
							break;
						}
						// left pointer is not empty
						else {
							current = current->left;
						}
					}
					else {
						// right pointer is empty
						if (current->right == NULL) {
							current->right = makeNode(data);
							break;
						}
						// right pointer is not empty
						else {
							current = current->right;
						}
					}
				}
			}

}
bNode *makeNode(char *data){
	// Function that allocates a bNode
	bNode* node = (bNode*) malloc (sizeof(bNode));
	node->data = data;
	node->left = NULL;
	node->right = NULL;
	return node;
}

int keyCompareToIgnoreCase(char *name1, char *name2){
	// Modified version of strcasecmp to handle proper names.
	// get last names
	char *lastname1 = lastToken(name1);
	char *lastname2 = lastToken(name2);
	// compare last names
	return strcasecmp (lastname1,lastname2);
}
char *lastToken(char *input){
	// Returns index of last token in string.
	for ( int i = strlen(input)-1; i >= 0; i--){
		  char c = input[i];
		  if (c ==' '){
			  // input points to 1st char
			  // input+i points to space
			  // input+i+1 points to 1st letter of the last name
			  return (char*) (input+i+1);
		  }
	}
	// if we reach this point, there is no space in the input string
	return input; // no space, compare with entire input
}
void displayInOrder(){
	// Prints bTree in forward order by calling recursive traversal function
	traverseInOrder(root);
}
void traverseInOrder(bNode *root){
	// left, root, right
	if (root->left != NULL){
		traverseInOrder (root ->left);
	}
	// visiting
	printf("%s\n",root->data);
	if (root->right != NULL){
		traverseInOrder(root->right);
	}
	// The recursive function that performs in order traversal
}

void displayInReverseOrder(){
	traverseInReverseOrder(root);
}
void traverseInReverseOrder(bNode *root){
	// right, root, left
	if (root->right!= NULL){
		traverseInReverseOrder (root ->right);
		}
	printf("%s\n",root->data);
	if (root->left != NULL){
		traverseInReverseOrder(root->left);
		}
}


//deallocation methods:
void deleteTree(){
	// Post order traversal used to delete tree nodes
	postOrder(root);
}

void postOrder (bNode *root){
	// Post order traversal used to delete tree nodes
	if (root->left!= NULL){
		postOrder (root ->left);
		}
	if (root->right != NULL){
		postOrder(root->right);
		}
	// postOrder
	free(root->data);
	free(root);
	}


int getNumNodes (){
	return numNodes;
}

File: A2/bTree.java
/** Woojin Jang (Student #260762536)
* this code is from lecture note created by professor frank p. ferrie
*/
public class bTree {

	bNode root;
	
	// bNode nested class 
	class bNode {
		String data;
		bNode left, right;
		
		// constructor of bNode
		public bNode(String data) {
			this.data = data;
		}
	}
	
	/** this code is from lecture note created by professor frank p. ferrie
	 */
	public void addNode(String data) {
		bNode current;
		// empty tree
		if (root == null) {
			this.root = new bNode(data);
		} 
		// not empty
		else {
			current = this.root;
			while (true) {
				// the return value is negative 
				// which means the name (data) comes first compare to the root
				if (keyCompareToIgnoreCase(data, current.data) < 0) {
					// left pointer is empty
					if (current.left == null) {
						current.left = new bNode(data);
						break;
					} 
					// left pointer is not empty
					else {
						current = current.left;
					}
				} 
				else {
					// right pointer is empty
					if (current.right == null) {
						current.right = new bNode(data);
						break;
					}
					// right pointer is not empty
					else {
						current = current.right;
					}
				}
			}
		}
	}

	/**
	 *  this code is from lecture note by professor frank p. ferrie
	 */
	// inorder left root right
	public void displayInOrder(bNode current) {
		// try catch is from geeks&geeks website
		try {
			if (current.left != null)
				displayInOrder(current.left);
			System.out.println(current.data);
			if (current.right != null)
				displayInOrder(current.right);
		}
		// if there is no input it's nullpointerexception
		catch (NullPointerException e) {
			System.out.println("NullPointerException");
		}
		
	}

	/**
	 *  this code is from lecture note by professor frank p. ferrie
	 */
	// reverse inorder right root left
	public void displayInReverseOrder(bNode current) {
		try {
			if (current.right != null)
				displayInReverseOrder(current.right);
			System.out.println(current.data);
			if (current.left != null)
				displayInReverseOrder(current.left);
		}
		// if there is no input it's nullpointerexception
		catch(NullPointerException e) {
			System.out.println("NullPointerException");
		}
	}
	/**
	 *  this code is provided by Katrina Pouline 
	 */
	// compare 2 last name to compare
	public int keyCompareToIgnoreCase(String name1, String name2) {
		// getting an empty index before last name
		int lastIndex1 = name1.lastIndexOf(" ");
		int lastIndex2 = name2.lastIndexOf(" ");
		// put last name index to last name
		// lastIndex contains a space, to ignore that need to add 1
		String lastName1 = name1.substring(lastIndex1 + 1); 
		String lastName2 = name2.substring(lastIndex2 + 1);
		// return if the lastName 2 goes to left (<0) or right (>0)
		return lastName1.compareToIgnoreCase(lastName2);
	}
}
